Update documentation for supported token-exchange (#38008)
keycloak/keycloak: docs/guides/securing-apps/token-exchange.adoc

integration might be available to all grants. It will not be specific only to token exchange as it was in token exchange V1.

* Integration token exchange with link:{adminguide_link}#_client_policies[Client policies] is possible. This can be useful to address the use cases like for example: Reject the token exchange request if
client `requester-client` send the request with `scope=some-confidential-scope`. For that particular example, it can be useful to create client policy condition with combined conditions
for `client-scope`, `grant-type` and `client-roles`.

* Requesting refresh token is allowed just if client has switch `Allow refresh token in Standard Token Exchange` set to other value than `No` (which is the default value). The switch is available in the
admin console in the tab `Advanced` of the OIDC client. The other available value of the switch is `Same session`, which means that refresh token is allowed just if refresh token can use the same user
session like the subject token has. In case that subject token is coming from transient session or from link:{adminguide_link}#_offline-access[Offline session], the requesting refresh token will not

> be allowed. Similarly it will not be allowed to request offline token (using `scope=offline_access`).

Reviewer: Maybe we can add another image with the configuration option. Saying the the option is in the client `Advanced` tab in the section `OpenID Connect Compatibility Modes`.

Author: Good catch. Updated the PR with the image added.

* For the case when `access-token1` was exchanged to `refresh-token2` of client `requester-client`, we try to support revocation chain. This means that:
  ** Revocation of `access-token1` will revoke also `refresh-token2`. Moreover this will remove the client session of the client `requester-client` from the
     user session and hence all refresh tokens of `requester-client` in this user session will be effectively revoked
  ** In case that `refresh-token2` and it's related access token was used for the further token exchange to different client, then revocation of `access-token1` will revoke those subsequent token exchanges
     as well. In other words, whole "chain" of exchanged tokens is going to be revoked.
  ** Note that access token should be valid when the revocation endpoint is invoked. If you do not have valid access token as the original `access-token1` is already expired, you can perhaps use other
     access token issued to same client in the same user session. The exchanged tokens like `refresh-token2` and others from the "chain" should be revoked.

[[_legacy-token-exchange]]
== Legacy token exchange

> 

Reviewer: We can add another note [here](https://github.com/keycloak/keycloak/blob/28dab5ef1046dd046a83a13c5be0c923f7679e63/docs/guides/securing-apps/token-exchange.adoc?plain=1#L300-L318) saying that now to use legacy internal to internal token exchange you need to disable the new standard token exchange: `--features token-exchange,admin-fine-grained-authz --features-disabled=token-exchange-standard`. For the migration period.

Author: Thanks for noticing this. It is not needed to disable `token-exchange-standard` as if the client does not have `Standard token exchange` switch enabled, it won't serve requests by `StandardTokenExchangeProvider` and will fallback to V1. So people using legacy V1 token exchange do not need to do anything if they still want to stick to V1. In other words, once we enable standard token exchange in `main`, I hope that `StandardTokenExchangeV1Test` will still work the same as before without any changes needed in that test.

But I've rather added this explicitly to the "migration" section as I did not mentioned it before :-) Also added some note to "internal-internal" of V1 token exchange and people should rather use the standard.

priority=120
summary="Configuring and using Token exchange with {project_name}">

> Token exchange is the process, which allows client application to exchange one token for another token. In {project_name}, there are two features, which implement token exchange:

Reviewer 2: ```suggestion
Token exchange is the process that allows a client application to exchange one token for another token. In {project_name}, two features implement token exchange:
```

* <<_standard-token-exchange,Standard token exchange: version 2 (V2)>> - This is fully supported token exchange implementation, which is enabled by default once {project_name} server is started.
* <<_legacy-token-exchange,Legacy token exchange: version 1 (V1)>> - This is preview feature, which is not enabled by default once {project_name} server is started.

> Here is a short summary of the current capabilities of {project_name} around token exchange.

Reviewer 2: ```suggestion
The capabilities of {project_name} for token exchange are as follows:
```

. A client can exchange an existing {project_name} token created for a specific client for a new token targeted to a different client in the same realm
. A client can exchange an existing {project_name} token for an external token, i.e. a linked Facebook account
. A client can exchange an external token for a {project_name} token.
. A client can impersonate a user

> The standard token exchange supports only the use-case (1), when legacy token exchange provides also other three use-cases. When possible, it is highly recommended to use standard token exchange V2 as that
> one is supported and will be maintained for the future. Using legacy token exchange is possible if you need some of the last three use cases, but note that legacy token exchange is preview and may not be
> backwards compatible with the future {project_name} versions. It is possible to enable both token exchange features and use them together, which is useful if you need to use both internal-internal exchange
> provided by V2 together with other use cases, which are provided only by V1. The summary comparison of standard and legacy token exchange can be seen in <<_standard-token-exchange-comparison,this section>>.

Reviewer 2: ```suggestion
The standard token exchange supports only use-case (1). The legacy token exchange support the four use-cases, but it is a preview feature. Therefore, the standard token exchange V2 is recommended  since it is supported and will be maintained for the future. The legacy token exchange is useful for last three use cases, but it may not be
backwards compatible with the future {project_name} versions. You can also enable both token exchange features and use them together. For example, you could use both internal-internal exchange
provided by V2 together with other use cases that are supported by V1. For more details, see this <<_standard-token-exchange-comparison,token exchange comparison>>.
```

[[_standard-token-exchange]]
== Standard token exchange

Standard token exchange in {project_name} implements the https://datatracker.ietf.org/doc/html/rfc8693[Token exchange specification]. It allows client application to exchange an existing {project_name} token created

> for a specific client for a new token issued to the client, which triggered the token exchange request. Both clients must be in the same realm.

Reviewer 2: ```suggestion
for a specific client for a new token issued to the client that triggered the token exchange request. Both clients must be in the same realm.
```

[[_standard-token-exchange-flow]]
=== Token exchange flow

Typical token exchange flow may look like this:

> . User authenticates with the use of the {project_name} SSO to some client application `initial-client`. The token will be issued to the `initial-client`.

Reviewer 2: ```suggestion
. The user authenticates with the use of the {project_name} SSO to the client application `initial-client`. The token is issued to the `initial-client`.
```

> . Client `initial-client` may need to use some REST service called `requester-client`, which requires authentication. So `initial-client` sends the access token from step 1 to the `requester-client` with the
> use of the token

Reviewer 2: ```suggestion
. The client `initial-client` may need to use the REST service `requester-client`, which requires authentication. So the `initial-client` sends the access token from step 1 to the `requester-client` with the
use of the token
```

> . To serve the request, the `requester-client` may need to call another service called `target-client`. However it possibly cannot use the token sent to it from `initial-client` as for example:

Reviewer 2: ```suggestion
. To serve the request, the `requester-client` may need to call another service `target-client`. However it possibly cannot use the token sent to it from `initial-client` as for example:
```

> * The token does not have enough permissions or scopes

Reviewer 2: ```suggestion
* The token has insufficient permissions or scopes
```

> * The `target-client` is not specified as the token audience as token was intended to be used to invoke `requester-client`
> * Token has too much permissions and hence `requester-client` does not want to share it with the `target-client`
> +

Reviewer 2: ```suggestion
* The `target-client` is not specified as the token audience as the token was intended to be used to invoke `requester-client`
* The token has excess permissions and therefore the `requester-client` does not want to share it with the `target-client`
+
```

> Any of those could be the reason to invoke the token exchange. The `requester-client` may need to send token exchange request to the {project_name} server and use the original token from step 1 as

Reviewer 2: ```suggestion
Any of these situations could be the reason to invoke the token exchange. The `requester-client` may need to send the token exchange request to the {project_name} server and use the original token from step 1 as the 
```

_subject token_ and exchange it for another token _requested token_.

> . The _requested token_ is returned to `requester-client`. This token can be used to be sent to the `target-client`.

Reviewer 2: ```suggestion
. The _requested token_ is returned to `requester-client`. This token can now be sent to the `target-client`.
```

> . The `target-client` is able to fulfill the request and returns the response to the `requester-client`. The `requester-client` can then follow and return the response to the request from step 2

Reviewer 2: ```suggestion
. The `target-client` can fulfill the request and return the response to the `requester-client`. The `requester-client` can then follow and return the response to the request from step 2
```

There can be multiple other use cases where token exchange can be useful, but the one above is the most typical.

==== Example token exchange request

> The example token exchange request of the client `requester-client` in the realm `test` can look like this. Note that `subject_token` is the access token issued to the `initial-client`:

Reviewer 2: ```suggestion
The folowing is an example token exchange request of the client `requester-client` in the realm `test`. Note that `subject_token` is the access token issued to the `initial-client`:
```

[source,bash]
----
POST /realms/test/protocol/openid-connect/token
Authorization: Basic cmVxdWVzdGVyLWNsaWVudDpwYXNzd29yZA==
Content-Type: application/x-www-form-urlencoded
Accept: application/json

grant_type=urn:ietf:params:oauth:grant-type:token-exchange&
subject_token=$SUBJECT_TOKEN&
subject_token_type=urn:ietf:params:oauth:token-type:access_token&
requested_token_type=urn:ietf:params:oauth:token-type:access_token
----

The example token exchange response may look like this:

[source,json]
----
{
  "access_token": "eyJhbGciOiJSUzI1NiIsIn...",
  "expires_in": 300,
  "token_type": "Bearer",
  "issued_token_type": "urn:ietf:params:oauth:token-type:access_token",
  "session_state": "287f3c57-32b8-4c0f-8b00-8c7db231d701",
  "scope": "default-scope1",
  "refresh_expires_in": 0,
  "not-before-policy": 0
}
----

[[_standard-token-exchange-enable]]
=== How to enable token exchange

> To enable the standard token exchange, it is needed to have feature `token-exchange-standard:v2` enabled, which is by default. Besides that, it is needed to enable `Standard token exchange` switch for

Reviewer 2: ```suggestion
For standard token exchange, `token-exchange-standard:v2` is enabled by default.  However, you also need to enable the *Standard token exchange* switch for
```

Reviewer 2: We use *asterisks* for what you see in the Admin Console

> the client, which is supposed to send token exchange requests. Like `requester-client` from <<_standard-token-exchange-flow,the previous example>>. Note that `requester-client` must be confidential client

Reviewer 2: ```suggestion
the client that is supposed to send token exchange requests, such as the `requester-client` from the <<_standard-token-exchange-flow,the previous example>>. Note that `requester-client` must be a confidential client
```

> and similarly like for the other grant requests, the token exchange requests must be authenticated by the appropriate link:{adminguide_link}#_client-credentials[client authentication method] configured
> for the client.

Reviewer 2: ```suggestion
Also, as is the case for other grant requests, the token exchange requests must be authenticated by the appropriate link:{adminguide_link}#_client-credentials[client authentication method] that is configured
for the client.
```

.Enabling Token Exchange
image::token-exchange-switch.png[Enabling Token Exchange]

[[_standard-token-exchange-request]]
=== Request and response parameters

The parameters are aligned with the https://datatracker.ietf.org/doc/html/rfc8693#name-token-exchange-request-and-[Token exchange specification]. Here some {project_name} specific details.

grant_type::
    _REQUIRED._ The value of the parameter must be `urn:ietf:params:oauth:grant-type:token-exchange`.

subject_token::
    _REQUIRED._ A security token that represents the identity of the party on behalf of whom the request is being made.

subject_token_type::

>     _REQUIRED._ This parameter is the type of the token passed in the `subject_token` parameter. This must be `urn:ietf:params:oauth:token-type:access_token` when standard token exchange is being used
>     as {project_name} does not support any other types for the standard token exchange.

Reviewer 2: ```suggestion
    _REQUIRED._ This parameter is the type of the token passed in the `subject_token` parameter. This must be `urn:ietf:params:oauth:token-type:access_token` when the standard token exchange is being used
    because {project_name} does not support other types for the standard token exchange.
```

requested_token_type::

>     _OPTIONAL._ This parameter represents the type of token the client wants to exchange for. Currently only oauth and OpenID Connect token types are supported. The default value for this

Reviewer 2: ```suggestion
    _OPTIONAL._ This parameter represents the type of token that the client wants to exchange for. In this version,  only oauth and OpenID Connect token types are supported. The default value for this
```

>     is `urn:ietf:params:oauth:token-type:access_token` . Other possible value is `urn:ietf:params:oauth:token-type:id_token` if ID token issued to `requester-client` is requested. The possible value might
>     be also `urn:ietf:params:oauth:token-type:refresh_token` in which case you will be returned both an access token and refresh token within the response. However refresh token is allowed just if the client
>     configuration option `Allow refresh token in Standard Token Exchange` is enabled as specified in <<_standard-token-exchange-details,this section>>

Reviewer 2: ```suggestion
    is `urn:ietf:params:oauth:token-type:access_token` . Other possible value is `urn:ietf:params:oauth:token-type:id_token` if ID token issued to `requester-client` is requested. The possible value might
    be also `urn:ietf:params:oauth:token-type:refresh_token`; in this case, you will receive both an access token and refresh token within the response. However, the refresh token is allowed if the `Allow refresh token in Standard Token Exchange` client configuration option  is enabled as specified in <<_standard-token-exchange-details, the standard token exchange>> section.
```

scope::

>    _OPTIONAL._ This parameter represents the space-delimited set of OAuth and OpenID Connect scopes the client is requesting. It is allowed to use link:{adminguide_link}#_client_scopes[Optional client scopes]

Reviewer 2: ```suggestion
   _OPTIONAL._ This parameter represents the space-delimited set of OAuth and OpenID Connect scopes that the client is requesting. You can use link:{adminguide_link}#_client_scopes[Optional client scopes]
```

>    of the `requester-client`. More details in <<_standard-token-exchange-scope,the section about scopes and audiences>>. Not using this parameter means that only
>    the link:{adminguide_link}#_client_scopes[Default client scopes] will be effectively used.

Reviewer 2: ```suggestion
   of the `requester-client`. For more details, see <<_standard-token-exchange-scope,scopes and audiences>>. Omitting this parameter means that only
   the link:{adminguide_link}#_client_scopes[Default client scopes] are effectively used.
```

audience::
  _OPTIONAL._ Audience specifies `client_id` of the client, which is supposed to be used as the token audience. In <<_standard-token-exchange-flow,the example above>>, it could be `target-client`. Multiple
  values of this parameter are allowed, which means that you want the token to contain multiple audiences to be used by `requester-client` in multiple different services. For example
  `audience=target-client1&audience=target-client2` can be used in the request. More details in <<_standard-token-exchange-scope,the section about scopes and audiences>>.

> Successful response is returned in the JSON format. It contains similar parameters like the response from other grants. The interesting parameters with some token exchange specifics are those:

Reviewer 2: ```suggestion
A successful response is returned in the JSON format. It contains similar parameters such as the response from other grants. The following are some token exchange specifics of the more notable parameters: 
```

access_token::
    The requested access token. Note that if request specified `requested_token_type=urn:ietf:params:oauth:token-type:id_token`, this parameter may actually contain the ID token instead of access token.

>     This behaviour is per https://datatracker.ietf.org/doc/html/rfc8693#section-2.2.1[the token exchange specification].

Reviewer 2: ```suggestion
    This behavior is per https://datatracker.ietf.org/doc/html/rfc8693#section-2.2.1[the token exchange specification].
```

refresh_token::
    The refresh token. It is included just if `requested_token_type=urn:ietf:params:oauth:token-type:refresh_token` is used and the client has enabled issuing refresh tokens from the token exchange

issued_token_type::
    The issued requested token type. Same value as `requested_token_type` used in the request.

token_type::
    Usually `Bearer` if issued token type was access token or refresh token. In case of ID token requested, the value is `N_A`

[[_standard-token-exchange-scope]]
=== Scopes and audiences

> Parameter `scope` in the token exchange request has same meaning like for any other grants. It is optional parameter. When it is not used, the effective client scopes used in the request will be
> the link:{adminguide_link}#_client_scopes[Default client scopes]  of the `requester-client`. When it is used, the effective client scopes will be default scopes together with
> the link:{adminguide_link}#_client_scopes[Optional client scopes]

Reviewer 2: ```suggestion
The `scope` parameter in the token exchange request has the same meaning as for other grants. This parameter is optional. When it is omitted, the effective client scopes used in the request are
the link:{adminguide_link}#_client_scopes[Default client scopes]  of the `requester-client`. When this parameter is used, the effective client scopes are the default scopes together with
the link:{adminguide_link}#_client_scopes[Optional client scopes]
```

By default, the used client scopes will add the audiences to the `aud` claim of the token based on the used client scopes and client roles as specified in the link:{adminguide_link}#audience-support[Audience documentation].

The `audience` parameter can be used for filtering of audiences, so that the `aud` claim will contain only the audiences specified by the `audience` parameter. Similarly the client roles in the token will

> be filtered and token will have only client roles of the clients specified by the `audience` parameter.

Reviewer 2: ```suggestion
be filtered and the token will have only the client roles of the clients specified by the `audience` parameter.
```

> Moreover, the `audience` parameter can be used to possibly filter client scopes as well. It works in a bit similar way to link:{adminguide_link}#client-scopes-permissions[Client scope permission for users].

Reviewer 2: ```suggestion
In addition, the `audience` parameter can be used to potentially filter client scopes as well. It works in a manner that is similar way to link:{adminguide_link}#client-scopes-permissions[Client scope permission for users].
```

> If the client scope does not contain any client roles (EG. does not contain any roles or contains only realm roles), there is no additional filtering of client scopes. However if client scope contains any

Reviewer 2: ```suggestion
If the client scope does not contain any client roles (for example, it contains zero roles or it contains only realm roles), no additional filtering occurs for client scopes. However if the client scope contains any
```

> client role mappings, it must include some client roles of the clients requested by the `audience` parameter. Composite roles are taken into the consideration. If the client scope does not contain any client

Reviewer 2: ```suggestion
client role mappings, it must include some client roles of the clients requested by the `audience` parameter. Composite roles are taken into the consideration. If the client scope does not contain any client
```

roles of the clients requested by the `audience`, the client scope will be filtered.

> NOTE: The `audience` parameter can be used to filter the audiences, which are coming from used client scopes. But this parameter will not add additional audiences. When audience parameter is not used, there is

Reviewer 2: ```suggestion
NOTE: The `audience` parameter can be used to filter the audiences that are coming from the used client scopes. However, this parameter will not add more audiences. When the audience parameter is omitted, 
```

> no filtering. This means that parameter `audience` can be effectively used for "downscoping" the token to make sure that it contain just the requested audiences. On the other hand `scope` parameter is used
> to add optional client scopes and hence it can be used for "upscoping" and adding more scopes.

Reviewer 2: ```suggestion
no filtering occurs. As a result, the `audience` parameter is effectively used for "downscoping" the token to make sure that it contains only the requested audiences. However, the `scope` parameter is used
to add optional client scopes and hence it can be used for "upscoping" and adding more scopes.
```

==== Examples

> Here are some examples to better illustrate the behaviour for scopes and audiences.

Reviewer 2: ```suggestion
Here are some examples to better illustrate the behavior for scopes and audiences.
```

Assume we have the realm with:

* Client `target-client1` with the client role `target-client1-role`

* Client `target-client2` with the client role `target-client2-role`

* Client `target-client3` with the client role `target-client3-role`

* Client scope `default-scope1`. This client scope has role scope mapping for the client role `target-client1/target-client1-role`

* Client scope `optional-scope2`. This client scope has role scope mapping for the client role `target-client2/target-client2-role`

* Client `requester-client`, which has client scope `default-scope1` added as default client scope and scope `optional-scope2` added as an optional client scope

* Authenticated user, who is member of both `target-client1-role` and `target-client2-role`

The settings above means that using scope `default-scope1` will add the audience `target-client1` to the token and using `optional-scope2` will add the audience `target-client2`. This is because of the
audience resolving described in the link:{adminguide_link}#_audience_resolve[Audience documentation].

===== Example 1

Token exchange request sent with `scope=optional-scope2` and without audience parameter:

> There will be no filtering of audience. The scopes and audiences will be resolved like for any other grants as described in the link:{adminguide_link}#_client_scopes[Client scopes] and

Reviewer 2: ```suggestion
There will be no filtering of audience. The scopes and audiences will be resolved as is the case for any other grants as described in the link:{adminguide_link}#_client_scopes[Client scopes] and
```

> link:{adminguide_link}#_audience_resolve[Audience documentation] sections. The response token will be like this (claims not interesting for this example omitted for brevity):

Reviewer 2: ```suggestion
link:{adminguide_link}#_audience_resolve[Audience documentation] sections. The response token will be similar to this (claims not interesting for this example omitted for brevity):
```

	"target-client1": {
  	  "roles": [ "target-client1-role" ]
	},
	"target-client2": {
  	  "roles": [ "target-client2-role" ]
	}
  },
  ...
}
----

===== Example 2

Token exchange request sent with `scope=optional-scope2` and with `audience=target-client2`

Same like previous example, but `target-client1` audience and client roles filtered due audience parameter was included, but only with this `target-client2` client. The client scope `default-scope1` will be
also filtered due it contains some client roles, but at the same time, it does not contain any client roles of requested audience client `target-client2`. So token would be like:

[source,json]
----
{
  "azp": "requester-client",
  "scope": "optional-scope2",
  "aud": [ "target-client2" ],
  "resource_access": {
    "target-client2": {
      "roles": [ "target-client2-role" ]
    }
  },
  ...
}
----

===== Example 3

Token exchange request sent with `scope=optional-scope2` and with `audience=target-client2&audience=target-client3`

The `target-client3` is not part of the token audience as user does not have any roles. So in this case, the request will be rejected as some of the requested audiences are not available.

> NOTE: As mentioned in the token exchange specification, it is good practice to downscope the token as much as possible and use only the audiences needed. Ideally only single audience. This increases the probability

Reviewer 2: ```suggestion
NOTE: As mentioned in the token exchange specification, it is good practice to downscope the token as much as possible and use only the audiences needed. Ideally use a single audience. This strategy increases the probability
```

that request will be allowed.

> NOTE: If you have more complex deployment with many various scopes and audiences, it can be tricky to model it in an appropriate way. It is advised to use the link:{adminguide_link}#_client_scopes_evaluate[Client scopes evaluate tab]

Reviewer 2: ```suggestion
NOTE: If you have a more complex deployment with many various scopes and audiences, it can be challenging to model it in an appropriate way. Consider using the link:{adminguide_link}#_client_scopes_evaluate[Client scopes evaluate tab]
```

> to test if the token looks as expected for given user and for given set of scopes and audiences.

Reviewer 2: ```suggestion
to test if the token looks as expected for the given user and for the given set of scopes and audiences.
```

[[_standard-token-exchange-details]]
=== Token exchange - Additional details

> Few more points to clarify the behaviour of token exchange.

Reviewer 2: ```suggestion
These additional points clarify the behavior of token exchange.
```

* It is not supported for public clients to send the token exchange requests. The V1 had some very limited support to public clients, when public client can exchange the token to itself with less scopes.
This use case can be replaced by refresh token grant.

> * The `subject_token` sent to the token exchange endpoint must have the requester client set as an audience in the `aud` claim. Otherwise the request would be rejected. The only exception is, if client

Reviewer 2: ```suggestion
* The `subject_token` sent to the token exchange endpoint must have the requester client set as an audience in the `aud` claim. Otherwise, the request would be rejected. The only exception is, if client
```

> exchanges his own token, which was issued to it. Exchanging to itself might be useful to downscope/upscope the token or filter unneeded token audiences etc.

Reviewer 2: ```suggestion
exchanges his own token, which was issued to it. Exchanging to itself might be useful to downscope/upscope the token or filter unneeded token audiences and so on.
```

* Consents - If requester client has `Consent required` switch enabled, the token exchange is allowed just if user already granted consent to all requested scopes

* link:{adminguide_link}#_fine_grain_permissions[Fine-grained admin permissions (FGAP)] are not needed for the standard token exchange. We plan to eventually integrate with FGAP for the future, but that
integration might be available to all grants. It will not be specific only to token exchange as it was in token exchange V1.

> * Integration token exchange with link:{adminguide_link}#_client_policies[Client policies] is possible. This can be useful to address the use cases like for example: Reject the token exchange request if

Reviewer 2: ```suggestion
* Integrating token exchange with link:{adminguide_link}#_client_policies[Client policies] is possible. This integration can be useful to address certain use cases. For example, consider the use case to reject the token exchange request if the 
```

> client `requester-client` send the request with `scope=some-confidential-scope`. For that particular example, it can be useful to create client policy condition with combined conditions
> for `client-scope`, `grant-type` and `client-roles`.

Reviewer 2: ```suggestion
client `requester-client` sends the request with `scope=some-confidential-scope`. In this example, it can be useful to create a client policy condition with combined conditions
for `client-scope`, `grant-type` and `client-roles`.
```

> * Requesting refresh token is allowed just if client has switch `Allow refresh token in Standard Token Exchange` set to other value than `No` (which is the default value). The switch is available in the

Reviewer 2: ```suggestion
* Requesting a refresh token is allowed only if the client has the switch *Allow refresh token in Standard Token Exchange* set to a value other than `No` (the default value). The switch is available in the
```

> admin console in the tab `Advanced` of the OIDC client in the section `OpenID Connect Compatibility Modes` . The other available value of the switch is `Same session`, which means that refresh token is
> allowed just if refresh token can use the same user session like the subject token has. In case that subject token is coming from link:{adminguide_link}#_transient-session[Transient session] or from

Reviewer 2: ```suggestion
Admin Console in the *Advanced* tab of the OIDC client in the *OpenID Connect Compatibility Modes* section. The other available value of the switch is *Same session*, which means that the refresh token is
allowed only if the refresh token can use the same user session as the subject token. If that subject token is coming from a link:{adminguide_link}#_transient-session[Transient session] or from an 
```

> link:{adminguide_link}#_offline-access[Offline session], the requesting refresh token will not be allowed. Similarly it will not be allowed to request offline token (using `scope=offline_access`).

Reviewer 2: ```suggestion
link:{adminguide_link}#_offline-access[Offline session], the requesting refresh token will not be allowed. Similarly it will not be allowed to request an offline token (using `scope=offline_access`).
```

.Enabling refresh token in Token Exchange
image::token-exchange-switch-refresh.png[Enabling refresh token in Token Exchange]

> * Token exchange never creates new link:{adminguide_link}#managing-user-sessions[user session]. In case that `requested_token_type` is refresh token, it may eventually create new client session in the user session

Reviewer 2: ```suggestion
* Token exchange never creates a new link:{adminguide_link}#managing-user-sessions[user session]. In case that `requested_token_type` is a refresh token, it may eventually create a new client session in the user session
```

> for the requester client (if client session was not yet created).

Reviewer 2: ```suggestion
for the requester client (if the client session was not yet created).
```

* {project_name} Token exchange does not yet have support for the `resource` parameter.

> * Token exchange specification mentions concepts of https://datatracker.ietf.org/doc/html/rfc8693#name-delegation-vs-impersonation[impersonation and delegation]. {project_name} has support for the
> impersonation use case, but not yet for the delegation use case.

Reviewer 2: ```suggestion
* The token exchange specification mentions the concepts of https://datatracker.ietf.org/doc/html/rfc8693#name-delegation-vs-impersonation[impersonation and delegation]. {project_name} has support for the
impersonation use case, but not yet for the delegation use case.
```

==== Revocation

> Assuming that there is subject token `access-token1` issued to client `initial-client`. Here are some considerations related to token revocation:

Reviewer 2: ```suggestion
Assuming that there is a subject token `access-token1` issued to the client `initial-client`, here are some considerations related to token revocation:
```

> * For the case when `access-token1` was exchanged to `access-token2` of client `requester-client`, the revocation of `access-token1` will not revoke `access-token2`. Supporting of "revocation chain" for access
> tokens would mean quite an overhead. So considering this, it is the responsibility of the administrator that access tokens are short-lived and will be revoked automatically after some time.

Reviewer 2: ```suggestion
* For the case when the `access-token1` was exchanged to the `access-token2` of the client `requester-client`, the revocation of the `access-token1` will not revoke `access-token2`. Supporting of a "revocation chain" for access
tokens would mean quite an overhead. So considering this,  the administrator must ensure that access tokens are short-lived and are revoked automatically after some time.
```

* For the case when `access-token1` was exchanged to `refresh-token2` of client `requester-client`, we try to support revocation chain. This means that:
  ** Revocation of `access-token1` will revoke also `refresh-token2`. Moreover this will remove the client session of the client `requester-client` from the
     user session and hence all refresh tokens of `requester-client` in this user session will be effectively revoked
  ** In case that `refresh-token2` and it's related access token was used for the further token exchange to different client, then revocation of `access-token1` will revoke those subsequent token exchanges

>      as well. In other words, whole "chain" of exchanged tokens is going to be revoked.

Reviewer 2: ```suggestion
     as well. In other words, the whole "chain" of exchanged tokens is going to be revoked.
```

>   ** Note that access token should be valid when the revocation endpoint is invoked. If you do not have valid access token as the original `access-token1` is already expired, you can perhaps use other

Reviewer 2: ```suggestion
  ** Note that the access token should be valid when the revocation endpoint is invoked. If you do not have a valid access token as the original `access-token1` is already expired, you can perhaps use another
```

>      access token issued to same client in the same user session. The exchanged tokens like `refresh-token2` and others from the "chain" should be revoked.

Reviewer 2: ```suggestion
     access token issued to same client in the same user session. The exchanged tokens such as `refresh-token2` and others from the "chain" should be revoked.
```

[[_standard-token-exchange-comparison]]
=== Comparison of standard token exchange and legacy token exchange

> Here is the summary with the comparison of current capabilities of standard token exchange and legacy token exchange. Details are described in the sections above.

Reviewer 2: ```suggestion
While the preceding sections fully detail standard and legacy token exchange, the following is an overall summary that compares the two token exchange methods. 
```

[cols="3*", options="header"]
|===
|Capability |Standard token exchange V2 |Legacy token exchange V1
s|Internal-internal token exchange | Supported. Implemented as per rfc8693 | Preview support. Loose implementation of rfc8693. It is recommended to use V2 instead
s|Allowed `subject_token_type`     | Access token type only   | Access token type only for internal-internal, JWT for external-internal scenarios
s|Allowed `requested_token_type`   | Access token (default), Refresh token, ID token | Access token, Refresh token (default), SAML2 assertion
s|Behaviour of `scope` parameter   | Aligned with other grants. Scope parameter means requesting optional scopes of the client, which sent the token exchange request  | Scope parameter based on scopes of
the "target" client specified by audience parameter. Downscoping support only

> s|Behaviour of `audience` parameter   | Support for more values as per the specification. Can be used to narrow down the available audiences and keep only the requested audiences. Effectively downscoping the token per

Reviewer 2: ```suggestion
s|Behavior of `audience` parameter   | Support for more values as per the specification. Can be used to narrow down the available audiences and keep only the requested audiences. Effectively downscoping the token per
```

the required target audience | Support for single audience value. Token effectively issued to the client requested by audience parameter and using scopes of that client
s|Public clients  | Not available. Downscoping implemented by V1 can be replaced by refresh token grant| Available only to exchange token of the client itself. Effectively downscoping support only

> s|Consents        | Allowed for clients with `Consent required` as long as user already granted consent | Not allowed for clients with `Consent required`

Reviewer 2: ```suggestion
s|Consents        | Allowed for clients with `Consent required` as long as the user is already granted consent | Not allowed for clients with *Consent required*
```

> s|Authorization  | Verification that requester client must be in the audience of `subject_token`. Integration with client policies. No Fine-grained admin permissions | Based on fine-grained admin permissions

Reviewer 2: ```suggestion
s|Authorization  | Verification that the requester client must be in the audience of the `subject_token`. Integration with client policies. No Fine-grained admin permissions | Based on fine-grained admin permissions
```
